### PR TITLE
vmware-horizon-client: init at 2006

### PR DESCRIPTION
**Pull request by buckley310**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jonringer** [comment]
@GrahamcOfBorg eval

**Comment by jonringer** [comment]
@GrahamcOfBorg eval

**Comment by das-g** [comment]
Result of `nixpkgs-review pr 96849` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
<br>- vmware-horizon-client
</details>

**Comment by das-g** [comment]
I can use the resulting `vmware-view` executable to connect to `vdi.hsr.ch` (which uses a SWIVEL 2nd factor) and start a desktop session there. However, when I try to connect to `vdi.rj.ost.ch` (which uses Microsoft login with mandatory 2nd factor), it fails with
```
2020-08-31 22:52:54.612+02:00: vmware-view 7937| udpProxyLib: FECHostSocket_Specialise: failed to get desired receive bufsize: 425984, errno 115

2020-08-31 22:52:54.612+02:00: vmware-view 7937| udpProxyLib: socket 102 transition to state SYN_SENT, reason FECSocketDoConnect refCount 4

2020-08-31 22:52:54.627+02:00: vmware-view 7937| udpProxyLib: socket 102 transition to state CLOSED, reason FECSocketDoClose refCount 6

2020-08-31 22:52:54.627+02:00: vmware-view 7937| udpProxyLib: FECSocketDestroy: Destroying listen socket 102

XPCOMGlueLoad error for file /nix/store/2xhs6qbip2j1sdhqpl0iaw11r4jfl32d-firefox-unwrapped-79.0/lib/firefox/libmozsandbox.so:
/usr/lib/vmware/gcc/libstdc++.so.6: version `GLIBCXX_3.4.26' not found (required by /nix/store/2xhs6qbip2j1sdhqpl0iaw11r4jfl32d-firefox-unwrapped-79.0/lib/firefox/libmozsandbox.so)
Couldn't load XPCOM.
2020-08-31 22:52:55.629+02:00: vmware-view 7937| TaskCombiner: Removing left CdkGetLaunchItemsTask(DEAD) from combiner.
2020-08-31 22:52:55.629+02:00: vmware-view 7937| TaskCombiner: Removing left CdkGetTunnelConnectionTask(DEAD) from combiner.
2020-08-31 22:52:55.629+02:00: vmware-view 7937| TaskCombiner: Removing left CdkGetTunnelConnectionTask(DEAD) from combiner.
2020-08-31 22:52:55.629+02:00: vmware-view 7937| TaskCombiner: Removing left CdkGetUserGlobalPreferencesTask(DEAD) from combiner.
2020-08-31 22:52:55.629+02:00: vmware-view 7937| TaskCombiner: Removing left CdkGetUserGlobalPreferencesTask(DEAD) from combiner.
2020-08-31 22:52:55.629+02:00: vmware-view 7937| TaskCombiner: Removing left CdkGetLaunchItemsTask(DEAD) from combiner.
2020-08-31 22:52:55.629+02:00: vmware-view 7937| TaskCombiner: Combiner have 6 items left when it is freed.
```

(This is within the session launched by `nixpkgs-review pr 96849`. Maybe it works better when properly installed?)

**Comment by buckley310** [comment]
I get the same issue when connecting to that server, and my system is running off this branch right now.
Looks like in that case it's trying to launch firefox to handle authentication and firefox doesn't have what it needs inside the FhsUserEnv... I could just add firefox as a dependency, but I'm not sure if running firefox in there will cause trouble (primarily what happens if someone starts using that copy of firefox without knowing it was launched weird). I will give this issue a good look when I get a minute.

**Comment by buckley310** [comment]
Even after I make firefox happy, and horizon launches firefox successfully, horizon still crashes after. Not sure why at this point.

**Comment by buckley310** [comment]
Actually, when I tested this on Ubuntu, the client also exited after launching the browser, so I'm thinking this is intended behavior. I pushed the fix that makes firefox run properly, but beyond that I cannot test or debug the login process, as I do not have access to any SAML-enabled horizon environment. Can you see what happens now?

**Comment by fabianhauser** [comment]
I checked and the behaviour that vmware-view exits is intended - after successful login, the browser redirects to the url `vmware-view://vdi.rj.ost.ch/?useExisting=false&SAMLart=some-private-token-which-i-removed-here`, which opens `vmware-view` on Ubuntu. I suppose this url should be registered with xdg somehow?

**Comment by buckley310** [comment]
I added the scheme handler, and now the `vmware-view://` links activate the vmware client for me.

**Review by das-g — APPROVED** [review]

**Comment by das-g** [comment]
Result of `nixpkgs-review pr 96849` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
<br>- vmware-horizon-client
</details>

**Review by fabianhauser — APPROVED** [review]

**Comment by buckley310** [comment]
Just rebased on 441a7da, no code changes.

**Comment by buckley310** [comment]
Replaced some package name aliases with corresponding real names. Should pass ofborg tests now.

File: pkgs/applications/networking/remote/vmware-horizon-client/default.nix [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    buildInputs = [ makeWrapper ];
+    installPhase = ''
+      mkdir ext $out
+      find ${sysArch} -type f -print0 | xargs -0n1 tar -Cext --strip-components=1 -xf
+      mv ext/bin ext/lib ext/share "$out"/
+
+      rm -f "$out/lib/vmware/gcc/libstdc++.so.6"
+
+      makeWrapper "$out/bin/vmware-view" "$out/bin/vmware-view_wrapper" \
+          --set GTK_THEME Adwaita \
```
Comment (kevincox) [inline pkgs/applications/networking/remote/vmware-horizon-client/default.nix:32]: Why is this required? If possible could you link a public bug report or some other documentation of this issue? Or at least a comment?

File: pkgs/applications/networking/remote/vmware-horizon-client/default.nix [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+          --suffix LD_LIBRARY_PATH : "$out/lib/vmware/view/crtbora:$out/lib/vmware"
+    '';
+  };
+
+  desktopItem = makeDesktopItem {
+    name = "vmware-view";
+    desktopName = "VMware Horizon Client";
+    icon = "${vmwareHorizonClientFiles}/share/icons/vmware-view.png";
+    exec = "vmware-view %u";
```
Comment (kevincox) [inline pkgs/applications/networking/remote/vmware-horizon-client/default.nix:41]: I think this should be an absolute path.

File: pkgs/applications/networking/remote/vmware-horizon-client/default.nix (hunk 1 above)
Comment (buckley310) [reply pkgs/applications/networking/remote/vmware-horizon-client/default.nix:38]: I added comments describing why the two workarounds are present.

The library deletion is because the launch script tries to use its own bundled version of libstdc++ if it does not detect a new enough version already present. Frankly, the version check is a little hacky, and it's distro-specific and fails on NixOS. The bundled version causes issues, and ours works fine.

As for the theme issue, basically its pretty easy to screw up the UI by changing themes, even Adwaita-dark: https://communities.vmware.com/thread/623875

File: pkgs/applications/networking/remote/vmware-horizon-client/default.nix (hunk 2 above)
Comment (buckley310) [reply pkgs/applications/networking/remote/vmware-horizon-client/default.nix]: This is now an absolute path.
